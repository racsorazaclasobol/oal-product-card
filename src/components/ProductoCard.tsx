import React, { createContext } from 'react';

import { useProduct } from '../hooks/useProduct';
import { ProductContextProps, Product, onChangeArgs, InitialValues, ProductCardHandlers } from '../interfaces/productsInterfaces';

import '../styles/styles.css';

export const ProductContext = createContext({  } as ProductContextProps);
const { Provider } = ProductContext;

export interface ProductCardProps {
    className?      : string;
    counterValue?   : number;
    initialValues?  : InitialValues;
    product         : Product;
    style?          : React.CSSProperties;
    
    children        : ( args: ProductCardHandlers ) => JSX.Element;
    onChange?       : ( args: onChangeArgs ) => void;
}

export const ProductCard = ({ children, product, className, style, onChange, counterValue, initialValues }: ProductCardProps) => {
    
    const { counter, maxCounter, isMaxCounterReached, increaseBy, reset } = useProduct({ onChange, product, counterValue, initialValues });

    return (
        <Provider value={{ counter, maxCounter, increaseBy, product }}>
            <div className={`productCard ${ className  }`} style={ style } >

                {
                    children({
                        counter,
                        isMaxCounterReached,
                        maxCounter: initialValues?.maxCounter,
                        product,
                        
                        increaseBy,
                        reset
                    })
                }

            </div>
        </Provider>
    )
}