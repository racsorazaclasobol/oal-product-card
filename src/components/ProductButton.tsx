import React, { useCallback, useContext } from "react";
import { ProductContext } from "./ProductoCard";

export interface ProductButtonProps {
    className?  :string;
    style?      : React.CSSProperties;
}

export const ProductButton = ({ className }: ProductButtonProps) => {
    
    const { increaseBy, counter, maxCounter } = useContext( ProductContext );

    const isMaxReached = useCallback(() =>  !!maxCounter && counter === maxCounter, [counter, maxCounter],);
    
    
    return (
        <div  className={`buttonsContainer ${ className }`}>

            <button className='buttonMinus' onClick={ () => increaseBy( -1 ) }>
                -
            </button>

            <div className='countLabel'> 
                { counter } 
            </div>

            <button className={`buttonAdd ${ isMaxReached() && 'disabled' }`} onClick={ () => increaseBy( 1 ) } >
                +
            </button>
            
        </div>
    )
}