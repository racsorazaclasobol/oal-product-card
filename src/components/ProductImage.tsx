import React, { useContext } from "react";
import { ProductContext } from "./ProductoCard";
import noImage from '../assets/no-image.jpg'

export interface ProductImageProps {
    img?        : string;
    className?  : string;
    style?      : React.CSSProperties;
}

export const ProductImage = ( { img, className }: ProductImageProps ) => {
    
    const { product } = useContext( ProductContext );
    let imgToShow: string;

    imgToShow = ( img ) ? img : ( ( product?.img ) ? product.img : noImage );

    return (

        <img className={ `productImg ${ className }` } src={ imgToShow } alt='Product Image' />   
         
    )
}