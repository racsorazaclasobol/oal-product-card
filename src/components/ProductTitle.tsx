import React, { useContext } from "react";
import { ProductContext } from "./ProductoCard";

export interface TitleProps {
    title?: string;
    className?: string
    style?: React.CSSProperties;
}

export const ProductTitle = ({ title, className }: TitleProps) => { //{ title: string } simula ser una interfaz

    const { product } = useContext( ProductContext );

    let titleToShow: string; 
    titleToShow = ( !title ) ? product.title : title;

    return (
        <span className={`productDescription ${ className }`}> { titleToShow } </span>
    )
}