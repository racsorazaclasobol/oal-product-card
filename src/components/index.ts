import { ProductCardHOCProps } from '../interfaces/productsInterfaces';
import { ProductCard as ProductCardHOC }    from './ProductoCard';
import { ProductButton }    from './ProductButton';
import { ProductImage }     from './ProductImage';
import { ProductTitle }     from './ProductTitle';

export { ProductButton }    from './ProductButton';
export { ProductImage }     from './ProductImage';
export { ProductTitle }     from './ProductTitle';



export const ProductCard: ProductCardHOCProps = Object.assign( ProductCardHOC, {
    Title: ProductTitle,
    Image: ProductImage,
    Button: ProductButton,
} );

export default ProductCard;