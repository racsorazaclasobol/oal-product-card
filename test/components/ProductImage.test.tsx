import React from 'react';
import rendered from 'react-test-renderer';
import { ProductImage, ProductCard } from '../../src/components';
import { product2 } from '../data/products';


describe('Pruebas en ProductImage', () => { 
    
    test('Debe de mostrar el componene correctamene con la imagen personalizada', () => { 
        
        const wrapper = rendered.create(
            <ProductImage img={ product2.img } className='test-class' />
        )

        expect( wrapper.toJSON() ).toMatchSnapshot();

    });

    test('Debe de mostrar el componene correctamene con la imagen del producto', () => { 
        
        const wrapper = rendered.create(
            <ProductCard product={ product2 } >
                {
                    () => (
                        <ProductImage />
                    )
                }
            </ProductCard>
        )

        expect( wrapper.toJSON() ).toMatchSnapshot();

    })

})