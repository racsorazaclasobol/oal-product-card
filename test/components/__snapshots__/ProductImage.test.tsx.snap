// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en ProductImage Debe de mostrar el componene correctamene con la imagen del producto 1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="Product Image"
    className="productImg undefined"
    src="./coffee-mug.png"
  />
</div>
`;

exports[`Pruebas en ProductImage Debe de mostrar el componene correctamene con la imagen personalizada 1`] = `
<img
  alt="Product Image"
  className="productImg test-class"
  src="./coffee-mug.png"
/>
`;
